cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -0,0 +1,234 @@
+                                          @Assisted Set<Integer> partitions,
+                                          MessagingService messagingService,
+                                          SchemaGenerator schemaGenerator,
+                                          DatumReaderFactory readerFactory,
+                                          MetricStore metricStore,
+                                          @Named(Constants.Metrics.MESSAGING_FETCHER_PERSIST_THRESHOLD)
+                                          int persistThreshold) {
+    super(metricDatasetFactory);
+    this.metricsTopic = NamespaceId.SYSTEM.topic(topicPrefix);
+    this.metricsTopicMetaKey = new TopicIdMetaKey(this.metricsTopic);
+    this.partitions = partitions;
+    this.messagingService = messagingService;
+    try {
+      this.recordSchema = schemaGenerator.generate(MetricValues.class);
+      this.recordReader = readerFactory.create(TypeToken.of(MetricValues.class), recordSchema);
+    } catch (UnsupportedTypeException e) {
+      throw Throwables.propagate(e);
+    }
+    this.metricStore = metricStore;
+    this.persistThreshold = persistThreshold;
+    this.messageCount = new AtomicInteger();
+    this.lastMessageId = new AtomicLong(INITIAL_LAST_MESSAGE_ID);
+    this.metricsContextMap = Collections.<String, String>emptyMap();
+  }
+
+  @Override
+  public void setMetricsContext(MetricsContext metricsContext) {
+    this.metricsContext = metricsContext;
+    this.metricsContextMap = metricsContext.getTags();
+  }
+
+  @Override
+  protected void run() {
+    MetricsConsumerMetaTable metaTable = getMetaTable(this.getClass());
+    if (metaTable == null) {
+      LOG.info("Could not get MetricsConsumerMetaTable, seems like we are being shut down");
+      return;
+    }
+    try {
+      long messageId = metaTable.get(metricsTopicMetaKey);
+      MessageFetcher fetcher = messagingService.prepareFetch(metricsTopic);
+      if (messageId == -1L) {
+        // If no messageId is found for the last processed message, start fetching from beginning
+        fetcher.setStartTime(0L);
+      } else {
+        fetcher.setStartMessage(Bytes.toBytes(messageId), false);
+      }
+      while (isRunning()) {
+        processMetricsRecords(fetcher);
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to process metrics records", e);
+    }
+  }
+
+  private void processMetricsRecords(MessageFetcher fetcher) {
+    // Decode the metrics records.
+    ByteBufferInputStream is = new ByteBufferInputStream(null);
+    List<MetricValues> records = Lists.newArrayList();
+
+    try (CloseableIterator<RawMessage> iterator = fetcher.fetch()) {
+      while (iterator.hasNext()) {
+        RawMessage input = iterator.next();
+        try {
+          MetricValues metricValues = recordReader.read(
+            new BinaryDecoder(is.reset(ByteBuffer.wrap(input.getPayload()))), recordSchema);
```

Comment:
Don't create a new `BinaryDecoder` per event. Also, use a subclass of `ByteArrayInputStream` instead so that you reset the buffer as `byte[]` directly, avoiding the creation of the wrapping `ByteBuffer` object.